catalog/core/catalog-core-migratable/src/main/java/org/codice/ddf/catalog/migratable/impl/MigrationTaskManager.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.migratable.impl;
+
+import java.io.File;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.codice.ddf.migration.MigrationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import ddf.catalog.data.Result;
+
+/**
+ * Encapsulates all threading state for the catalog migration process and allows inspection of
+ * its progress through a public API to make testing easier.
+ */
+class MigrationTaskManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MigrationTaskManager.class);
+
+    private final ListeningExecutorService taskExecutor;
+
+    private final AtomicReference<Throwable> taskFailure;
+
+    private final MigrationFileWriter fileWriter;
+
+    private final CatalogMigratableConfig catalogConfig;
+
+    /**
+     * The creation of a task manager requires the passing of migratable configurations so that
+     * task behavior can be explicitly defined (i.e. where to export, how many metacards per file,
+     * and other useful data).
+     *
+     * @param config The set of configuration options to use.
+     */
+    public MigrationTaskManager(final CatalogMigratableConfig config,
+            final MigrationFileWriter fileWriter, final ExecutorService executor) {
+        this.taskExecutor = MoreExecutors.listeningDecorator(executor);
+        this.taskFailure = createAtomicReference();
+        this.catalogConfig = config;
+        this.fileWriter = fileWriter;
+    }
+
+    AtomicReference<Throwable> createAtomicReference() {
+        return new AtomicReference<>(null);
```

Comment:
We should also move that method after the public ones and add an inline comment saying that it is package-private for unit testing purposes.